### PR TITLE
Bezier-rs: Add parametric evaluate and line intersect to subpath

### PR DESCRIPTION
**Pull request by Androxium**
This PR adds the following
- `parametric_evaluate` to subpath
- `intersection` to subpath

This PR changes the following
- how cubic equations are solved due to a bug from the following edge case: incorrect root found when a handle and anchor point overlap on cubic curves

TODO: go back and fix failing unit tests caused by the bug fix

### CODE REVIEW

**Review by RobNadal — COMMENTED** [review]
I ran this locally as well and its working great - no flickering like before. Awesome stuff. LGTM

File: libraries/bezier-rs/src/subpath/solvers.rs [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+				let number_of_curves = self.len_segments() as f64;
+				let scaled_t = t * number_of_curves;
+
+				let target_curve_index = scaled_t.floor() as i32;
+				let target_curve_t = scaled_t % 1.;
+
+				if let Some(curve) = self.iter().nth(target_curve_index as usize) {
+					curve.evaluate(target_curve_t)
+				} else {
+					self.iter().last().unwrap().evaluate(1.)
```
[the remark this answers is not in the document]
Comment (hannahli2010) [reply libraries/bezier-rs/src/subpath/solvers.rs:22]: The `nth` function will return the nth element of the iterator, with indexing starting at 0. For any `t < 1`, the `target_curve_index` would round down to something `< number_of_curves` and hence actually return a curve when `nth` is called. This clause handles the case where `t = 1.` (aka return `t = 1` for the very last curve in the subpath).

[the remark this answers is not in the document]
Comment (Androxium) [reply libraries/bezier-rs/src/subpath/solvers.rs:22]: Yup, what happens is that when `t=1` for a bezier segment at index `i`, it will actually return the point `t=0` for the bezier segment `i+1`. This works well until we reach the last curve, in which case we would index out of bounds, that's what the else clause is checking. When `t=1` for the subpath, we want `t=1` for the last bezier segment.

**Comment by Keavon** [comment]
I hope we can improve the accuracy to avoid duplicate results at near-parallel angles:

https://user-images.githubusercontent.com/4388688/202835044-b14d8f90-0a37-42c8-b351-e11eb88cfba8.mp4

File: website/other/bezier-rs-demos/src/App.vue [hunk 2]
```diff
@@ -558,6 +566,38 @@ export default defineComponent({
 					name: "Length",
 					callback: (subpath: WasmSubpathInstance): string => subpath.length(),
 				},
+				{
+					name: "Evaluate",
+					callback: (subpath: WasmSubpathInstance, options: Record<string, number>): string => subpath.evaluate(options.t),
+					sliderOptions: [tSliderOptions], // TODO: add slider to toggle between parametric and euclidean
```
Comment (Keavon) [inline website/other/bezier-rs-demos/src/App.vue:572]: A few things:
- This should not be called a slider, call it either a checkbox or toggle. (A slider means a [range input](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/range))
- But when I merged Hannah's parametric/euclidean PR, I replaced that checkbox/toggle input with a radio button, please see how it looks now in the Evaluate function at <https://graphite.rs/bezier-rs-demos/>
- If the intention (as I'm assuming based on glancing at this code) is to separately implement the parametric/euclidean toggle, please directly copy the pertinent code from those Vue components now that I changed it to a pair of radio buttons.
    - The goal in the near future is to combine `BezierExample.vue` and `SubpathExample.vue`, as well as `BezierExamplePane.vue` and `SubpathExamplePane.vue`, into one component since that is duplicated code that makes maintenance (like this kind of dual implementation) more burdensome. I hope you guys can aim to put that goal on the roadmap for the relatively near future.
    - That'll be the first step towards removing Vue entirely. The next step will be merging `*ExamplePane.vue` into `App.vue`, and then `*Example.vue` also into `App.vue` so that becomes the only Vue component, prior to fully removing Vue.

So in summary: please remove all the code related to "slider options" (since that's the wrong terminology, but also because you'll want to directly copy the radio button-based implementation). Thanks! Also if you can get this done before Monday night, I will be able to review it before I leave early vacation Tuesday morning.

**Review by Keavon — CHANGES_REQUESTED** [review]
After the change I requested, the rest is good to go in terms of code quality, since I read over it all in my code review pass. However, fixing the TODO stuff (like the disabled tests) would be an important step, potentially before merging but later could also be ok. Also the accuracy issue I pointed out in the video I commented with.

**Comment by Keavon** [comment]
@Androxium just checking in on the status of this...

**Comment by Androxium** [comment]
@Keavon this PR is good to review

**Comment by Keavon** [comment]
This is failing CI, could you please fix that?

**Review by Keavon — APPROVED** [review]
For the record, we still have a few precision issues leading to multiple outputs, but it's way better now and this is sufficient for now (it can be revisited if it causes problems in the future).

image
